Implement Blobstore

vedantroy opened the pull request:
Implement the blobstore library as seen [here](https://github.com/vedantroy/step-2020/blob/master/walkthroughs/week-4-libraries/blobstore/blobstore-walkthrough.md).

Url: https://vedantroy-step-2020.uc.r.appspot.com/

Note: The file upload works locally, but does not seem to work when deployed. Currently figuring out what is going on with that.

Note 2: I noticed a bit of duplication of constants in my code. Should I extract them into a constants class?

@sooniln @alexanderkuznetsov96

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java:
```diff
@@ -0,0 +1,106 @@
+private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+@Override
+protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
```
nit: please fix indentation throughout the file - most ide's should have some sort of autoindent type thingy

vedantroy replied:
Fixed

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java:
```diff
@@ -0,0 +1,106 @@
+@WebServlet("/files")
+public class FileHandlerServlet extends HttpServlet {
+
+private static final String CONTENT_TYPE_JSON = "application/json;";
```
let's break down your question about constants a little more. generally, i'm a bit suspicious of files like a hypothetical Constants.java. they end up serving as dumping grounds for pretty much every constant, and the reason that can be a bad thing is that they will all have the same visibility. pretty much every piece of code, whether a constant or whatever, you want to have the minimum visibility possible. to consider some examples:

PATH_HOME_PAGE - this seems like it should have pretty high visibility, any piece of code that wants to redirect to the home page or something needs this constant

MAX_COMMENTS - there should be no reason for anyone outside of the code dealing with comments to see this. if there is outside code that needs to see this it's a good indication that code is doing something weird or incorrect, and may need to be fixed.

KEY_IMAGE_BLOBKEY - same logic, only code dealing with getting/setting images should ever need to see this.

Immediately we can see that we might require at least 3 different visibilities here. Thus, I would say the simplest solution is to keep each in the file where they are needed. PATH_HOME_PAGE etc, could potentially be extracted out to some PathConstants.java, etc, if you so desired.

One last comment re CONTENT_TYPE_JSON. This is a slightly unusual constant, in that its value can never ever change. It's defined in god knows how many specs, and if it's value ever changed, half the internet would break. Constants like these are more acceptable to use directly as literals in code, since we know there is a 0% chance they could change. Constants like KEY_IMAGE_URL, we might want to change in the future, and it makes sense to have a central reference.

A final thought on duplication: It's easier to consolidate duplication than to unconsolidate something. If you're in doubt, it's usually better to err on the side of duplicating (constants for example). If you change your mind, it's easy to centralize. If you've already centralized however, and later decide you want to change that, you may find it harder.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java:
```diff
@@ -0,0 +1,106 @@
+private static final String CONTENT_TYPE_JSON = "application/json;";
+private static final String KEY_IMAGE_BLOBKEY = "blobKey";
+private static final String KEY_IMAGE_URL = "url";
+private static final String KIND_IMAGE = "image";
```
KIND_IMAGE_URLS?

vedantroy replied:
`KIND_IMAGE` is for entities of kind `"url"` is the name of a property in that entity. Not sure if that answers your question.

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java:
```diff
@@ -0,0 +1,106 @@
+     }
+
+     // The form only contains a single file input
+     BlobKey blobKey = blobKeys.get(0);
```
what if there were multiple files? you might only care about 1, which is fine, but should you cleanup the rest from the blobstore anyways?

sooniln replied:
it might be helpful to think of this in terms of security as well - without this you've given a malicious user the ability to essentially charge you for unlimited storage...

vedantroy replied:
Fixed

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java:
```diff
@@ -0,0 +1,106 @@
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions opts = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    String imageUrl;
```
I don't understand the logic here very well, why validate the url and get it's path, but on failure, just use it anyways? dropping everything except the path portion of the url seems like it could cause errors too?

vedantroy replied:
A lot of the code in this servlet comes from here: https://github.com/vedantroy/step-2020/blob/master/walkthroughs/week-4-libraries/blobstore/examples/hello-world-fetch/src/main/java/com/google/sps/servlets/FormHandlerServlet.java

The context is apparently I need to do some shenanigans so stuff works on both the dev and deployed server.

sooniln commented on portfolio/src/main/webapp/script.js:
```diff
@@ -51,10 +51,14 @@ const form = document.getElementById("max-comments-form");
-  if (newMaxComments !== maxComments && newMaxComments !== NaN) {
-    maxComments = parseInt(newMaxComments);
-    getComments(maxComments);
+  for (const [key, val] of data) {
```
might need a rebase or something?

vedantroy replied:
Hmm... When I run `git rebase master`, it says the current branch is up to date, even though this code is definitely out of date. Not sure what to do.

vedantroy replied:
Ok, rebased. This should be gone.

vedantroy commented:
@sooniln I addressed comments.
- I left the constants as is, since, from your explanation, there aren't that many constants that should be shared across files (really, just the paths)
- I added code to delete extra blob keys
- Formatted the code
- Rebased to remove the weird Javascript regression

sooniln commented on portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java:
```diff
@@ -0,0 +1,39 @@
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
```
it might be better to have this as a member variable. currently you are paying the cost of BlobstoreServiceFactory.getBlobstoreService() for every get request - if it's a member variable you will only pay the cost once as long as the server keeps this servlet alive.

vedantroy replied:
Fixed

vedantroy commented:
@sooniln I overhauled the code so that each comment was associated with an optional image. As a result a decent chunk of code is outdated.
